Ipn templates

cmidgley opened the pull request:
Adds support for setting IPN based on Jinja2 templates.  See #38 for background, and the updated README for feature details.

cmidgley committed:
Add --ipn feature to set IPN on new pats
- Add --ipn to cli
- When part created and --ipn set, set IPN to the SKU (PartImporter.create_manufacturer_part)

cmidgley committed:
Adjust help text

cmidgley committed:
Revert fix to update_parameter
- moving to a different branch for separate pull

cmidgley committed:
Clean up endline spaces

cmidgley committed:
Fix crash when importing parts and parameters don't validate
- pass 'name' instead of trying to deref 'parameter.name'

cmidgley committed:
Merge remote-tracking branch 'origin/set-IPN-on-new-parts' into dev

30350n committed:
Remove name argument from update_parameter again

cmidgley committed:
Merge remote-tracking branch 'origin/fix-crash-on-invalid-property-value' into dev

cmidgley committed:
Merge branch 'master' into dev

cmidgley committed:
Merge branch 'master' into dev

cmidgley committed:
Rework IPN to be based on templates with part-category inheritance

cmidgley committed:
Merge branch '30350n:master' into dev

30350n committed:
Format code, Rename ipn_template to ipn_format, Refactor IPNSetting

30350n committed:
Clarify --ipn help string

30350n commented:
Okay, so some general thoughts on the implementation:
- Is it really necessary to use `jinja2`?
  I'd highly prefer an approach that just uses the standard [`str.format(...)` function](https://docs.python.org/3/library/stdtypes.html#str.format).
- It's nice that you already added extensive documentation for the feature, but I think it could be cut down a little bit. I'd prefer one or two sentences in the `config.yaml` section, with a simple example. (Let's do this last though, once we've settled on a final implementation).
- One thing I don't like is that this runs pre parameter-normalization. This is pretty my fault though, because it currently just gets done in `setup_parameters(...)`. I'll refactor this on master by creating splitting this up into a `match_parameters(...)` function which adjusts the `parameters` on the `api_part` directly and will be run before committing the parameters to InvenTree. Then your `import_part_ipn` can be run inbetweeen those.

cmidgley committed:
Fix type error on IPNSetting enum (introduced in commit 68f40ac)

cmidgley committed:
Merge branch 'master' into ipn-templates

cmidgley committed:
Clean up readme for consistency in name change from ipn template to ipn format

cmidgley committed:
If no ipn format specified, ignore IPN

cmidgley commented:
Merged latest master, cleaned up some additional parts due to the name change to `ipn_format`, and fixed a couple of small bugs.

To your earlier questions:
* jinja2: I do need templates (not just format strings) for things such as inline conditionals
* pre-parameter normalization: I did not see the `match_parameters` on master, so no changes were made for that.

cmidgley committed:
Update README.md to fix a small typo

Co-authored-by: Jorropo <jorropo.pgm@gmail.com>
